aptos-move/framework/aptos-framework/sources/aptos_governance.move

```diff
@@ -294,28 +358,76 @@ module aptos_framework::aptos_governance {
         proposal_id: u64,
         should_pass: bool,
-    ) acquires ApprovedExecutionHashes, GovernanceEvents, VotingRecords {
+    ) acquires ApprovedExecutionHashes, GovernanceEvents, VotingRecords, PartialVotingProposals, VotingRecordsV2 {
+        vote_internal(voter, stake_pool, proposal_id, MAX_U64, should_pass);
+    }
+
+    /// Vote on proposal with `proposal_id` and specified voting power from `stake_pool`.
+    public entry fun partial_vote(
+        voter: &signer,
+        stake_pool: address,
+        proposal_id: u64,
+        voting_power: u64,
+        should_pass: bool,
+    ) acquires ApprovedExecutionHashes, GovernanceEvents, VotingRecords, PartialVotingProposals, VotingRecordsV2 {
+        vote_internal(voter, stake_pool, proposal_id, voting_power, should_pass);
+    }
+
+    /// Vote on proposal with `proposal_id` and specified voting_power from `stake_pool`.
+    /// If voting_power is more than all the left voting power of `stake_pool`, use all the left voting power.
+    fun vote_internal(
+        voter: &signer,
+        stake_pool: address,
+        proposal_id: u64,
+        voting_power: u64,
+        should_pass: bool,
+    ) acquires ApprovedExecutionHashes, GovernanceEvents, VotingRecords, PartialVotingProposals, VotingRecordsV2 {
         let voter_address = signer::address_of(voter);
         assert!(stake::get_delegated_voter(stake_pool) == voter_address, error::invalid_argument(ENOT_DELEGATED_VOTER));
 
+        let is_partial_voting_proposal = false;
+        if (features::partial_governance_voting_enabled()) {
+            assert!(exists<PartialVotingProposals>(@aptos_framework),
+                error::invalid_state(EPARTIAL_VOTING_NOT_INITIALIZED));
+            assert!(exists<VotingRecordsV2>(@aptos_framework),
+                error::invalid_state(EPARTIAL_VOTING_NOT_INITIALIZED));
+            let partial_voting_proposals = borrow_global<PartialVotingProposals>(@aptos_framework);
+            is_partial_voting_proposal = table::contains(&partial_voting_proposals.proposals, proposal_id);
+        };
+
         // Ensure the voter doesn't double vote with the same stake pool.
         let voting_records = borrow_global_mut<VotingRecords>(@aptos_framework);
         let record_key = RecordKey {
             stake_pool,
             proposal_id,
         };
-        assert!(
-            !table::contains(&voting_records.votes, record_key),
-            error::invalid_argument(EALREADY_VOTED));
-        table::add(&mut voting_records.votes, record_key, true);
+        if (!is_partial_voting_proposal) {
+            assert!(
+                !table::contains(&voting_records.votes, record_key),
+                error::invalid_argument(EALREADY_VOTED));
+            table::add(&mut voting_records.votes, record_key, true);
+        };
 
-        let voting_power = get_voting_power(stake_pool);
+        let staking_pool_voting_power = get_voting_power(stake_pool);
+        if (is_partial_voting_proposal) {
```

Comment:
This logic can be greatly simplified by always checking voting records (v1) first, then if not present, check voting records v2. Also we can update vote() to dual write to both voting records v1 and v2 to simplify removal in the future (then we can just simply delete voting records v1 write once there are no longer proposals that are on the previous records)